libs/arrow/src/main/java/org/opensearch/arrow/StreamIterator.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.arrow;
+
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.opensearch.common.annotation.ExperimentalApi;
+
+import java.io.Closeable;
+
+/**
+ * An iterator over stream.
+ */
+@ExperimentalApi
+public interface StreamIterator extends Closeable {
```

Comment:
Let's make it `implements Iterator<VectorSchemaRoot>` (so it will stream friendly etc)?

Reply:
I'm glad you pointed it out. I was initially trying it but the fact that `VectorSchemaRoot` will remain same, its just that its content will be modified with call to next(), so it didn't make a lot of sense to iterate over `VectorSchemaRoot`. 

Also, [FlightStream](https://github.com/apache/arrow/blob/main/java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightStream.java), which is bidirectional stream, on which this iterator is meant to be built upon ,can make `hasNext()` call heavy duty. The fact that producer will not only produce when consumer has consumed the current batch and has requested the next batch using `next()`, so checking hasNext() might end up populating the next batch of data. Even if we optimize `hasNext()` to not populate data, its still an additional call, which can be avoided using next() returning a boolean.

Reply:
> I'm glad you pointed it out. I was initially trying it but the fact that `VectorSchemaRoot` will remain same, its just that its content will be modified with call to next(), so it didn't make a lot of sense to iterate over `VectorSchemaRoot`.

Ohhh... it changes the semantics, may be we should not call this an iterator? Hard to come up with the right name without usage, may be you could add some examples  to javadoc (as you did for `StreamProducer`)? It will be clear how this abstraction is supposed to be used ...